Add code insights lang stats creation UI

vovakulikov opened the pull request:
Close https://github.com/sourcegraph/sourcegraph/issues/20481

The third of the three PRs for insights creation UI. 

1. [Creation UI for search based insight]( https://github.com/sourcegraph/sourcegraph/pull/20032). https://github.com/sourcegraph/sourcegraph/pull/20032
2. [Introduction page for creation UI](https://github.com/sourcegraph/sourcegraph/pull/20482). https://github.com/sourcegraph/sourcegraph/pull/20482
3. **This Pull Request**

This PR is about to add form for lang stats insights. You can find [design layouts here](https://www.figma.com/file/QfeID0EEYMEPdEv8pbGULB/Code-Insights-WIP?node-id=262%3A11)

**How to test**

1. Open dev tools and set **_enableCodeInsightCreationUI_** field in local storage to true.
2. Go to /insight/create-intro and click create lang stats insight.

<img width="973" alt="image" src="https://user-images.githubusercontent.com/18492575/116996432-0d43cc00-ace4-11eb-8fe8-a074d91525c7.png">

vovakulikov commented:
I've added [story for creation lang stats insight page](https://5f0f381c0e50750022dc6bf7-hdywqfhqhj.chromatic.com/?path=/story/web-insights-createlangstatsinsightpageprops--page) for better visual review @AlicjaSuska fyi

vovakulikov commented on client/web/src/insights/pages/creation/lang-stats-creation/CreationLangStatsInsightPage.tsx:
```diff
@@ -0,0 +1,129 @@
+
+                // TODO [VK] Change these settings when multi code insights stats
+                // will be supported in code stats insight extension
+                const newSettingsString = {
```
It turned out code stats insight extension doesn't support multi configs, therefore now we can't have more than one stats insight. We should change extension first and then we able to merge this PR.

Joelkw commented on client/web/src/insights/pages/creation/lang-stats-creation/CreationLangStatsInsightPage.tsx:
```diff
@@ -0,0 +1,129 @@
+                <h2>Set up new language usage insight</h2>
+
+                <p className="text-muted">
+                    Shows usage of languages in your repository based on number of lines of code.{' '}
```
```suggestion
                    Shows language usage in your repository based on number of lines of code.{' '}
```

Joelkw commented on client/web/src/insights/pages/creation/lang-stats-creation/components/lang-stats-insight-creation-form/LangStatsInsightCreationForm.tsx:
```diff
@@ -0,0 +1,150 @@
+
+import styles from './LangStatsInsightCreationForm.module.scss'
+
+const repositoriesFieldValidator = createRequiredValidator('Repositories is required field for code insight.')
```
```suggestion
const repositoriesFieldValidator = createRequiredValidator('Repositories is a required field for code insight.')
```

Joelkw commented on client/web/src/insights/pages/creation/lang-stats-creation/components/lang-stats-insight-creation-form/LangStatsInsightCreationForm.tsx:
```diff
@@ -0,0 +1,150 @@
+import styles from './LangStatsInsightCreationForm.module.scss'
+
+const repositoriesFieldValidator = createRequiredValidator('Repositories is required field for code insight.')
+const thresholdFieldValidator = createRequiredValidator('Threshold is required field for code insight.')
```
```suggestion
const thresholdFieldValidator = createRequiredValidator('Threshold is a required field for code insight.')
```

Joelkw commented on client/web/src/insights/pages/creation/lang-stats-creation/components/lang-stats-insight-creation-form/LangStatsInsightCreationForm.tsx:
```diff
@@ -0,0 +1,150 @@
+                max={100}
+                type="number"
+                title="Threshold of ‘Other’ category"
+                description="The threshold for grouping all other languages into an 'other' category"
```
```suggestion
                description="Languages with usage lower than the threshold are grouped into an 'other' category"
```

Joelkw commented on client/web/src/insights/pages/creation/lang-stats-creation/components/lang-stats-insight-creation-form/LangStatsInsightCreationForm.tsx:
```diff
@@ -0,0 +1,150 @@
+                required={true}
+                autoFocus={true}
+                title="Repository"
+                description="This insight is limited to one repository. You can set up muliple language usage charts for many repositories."
```
```suggestion
                description="This insight is limited to one repository. You can set up multiple language usage charts for analyzing other repositories."
```

Joelkw commented on client/web/src/insights/pages/creation/lang-stats-creation/components/lang-stats-insight-creation-form/LangStatsInsightCreationForm.tsx:
```diff
@@ -0,0 +1,150 @@
+    title: string
+    repository: string
+    threshold: number
+    visibility: 'personal' | 'organisation'
```
```suggestion
    visibility: 'personal' | 'organization'
```

Joelkw commented on client/web/src/insights/pages/creation/lang-stats-creation/components/lang-stats-insight-creation-form/LangStatsInsightCreationForm.tsx:
```diff
@@ -0,0 +1,150 @@
+            <FormGroup
+                name="visibility"
+                title="Visibility"
+                description="This insight will be visible only on your personal dashboard. It will not be show to other
```
```suggestion
                description="This insight will be visible only on your personal dashboard. It will not appear for other
```

Joelkw reviewed (approved):
Also noting that (may be intentional at this stage) it won't let me select "organization" for visibility – only personal. 

Meta-note, not blocking/may not be worth doing before the release – can we combine our shared form creation fields (visibility, title, etc) so if we want to make future edits to design or copy, we need only make them once? (Or is this already planned/already the case and the PR split is just why they felt separate).

vovakulikov commented:
@Joelkw Thanks for this review. I fixed all texts from you (descriptions and error messages). I didn't make commits from you just because I had moved files and now these suggestions are outdated (my bad), but I added these suggestions manually.

About visibility, this is a great idea I think. I will extract this component (and other components) in shared form fields.

vovakulikov commented:
@valerybugakov If you got a moment take a look at this. This form does more or less that form from the first PR about creation UI but for another type of insight - code lang stats insight.

valerybugakov commented on client/web/src/insights/pages/creation/lang-stats/LangStatsInsightCreationPage.module.scss:
```diff
@@ -0,0 +1,5 @@
+@import '../../../../../../branded/src/global-styles/breakpoints';
```
```suggestion
@import 'branded/src/global-styles/breakpoints';
```

valerybugakov commented on client/web/src/insights/pages/creation/search-insight/SearchInsightCreationPage.module.scss:
```diff
@@ -0,0 +1,5 @@
+@import '../../../../../../branded/src/global-styles/breakpoints';
```
```suggestion
@import 'branded/src/global-styles/breakpoints';
```

vovakulikov replied:
@valerybugakov I tried this and with @sourcegrpah/branded

But 
<img width="969" alt="image" src="https://user-images.githubusercontent.com/18492575/117497808-9e839e80-af81-11eb-8ba4-8613bfb54a84.png">

vovakulikov replied:
Hmm. that happens only in web-dev/server I believe this path is working fine with the main (old) webpack config

vovakulikov replied:
@valerybugakov It turned out this import doesn't work in code sandbox webpack config. I moved `getCssLoader` function to `webpack` utils in order to reuse this in sandbox webpack.config

valerybugakov commented on client/web/src/insights/pages/creation/lang-stats/LangStatsInsightCreationPage.tsx:
```diff
@@ -0,0 +1,132 @@
+    const { history, authenticatedUser, settingsCascade, platformContext } = props
+    const { getSubjectSettings, updateSubjectSettings } = useContext(InsightsApiContext)
+
+    const handleSubmit = useCallback<LangStatsInsightCreationFormProps['onSubmit']>(
```
Logic in this file looks similar to [SearchInsightCreationPage.tsx](https://github.com/sourcegraph/sourcegraph/blob/24f60d343fd966982e07642ed4ad9f41eeabd4c9/client/web/src/insights/pages/creation/search-insight/SearchInsightCreationPage.tsx#L47). Maybe we can reuse some parts?

vovakulikov replied:
I thought about this but decided to copy-paste it just to avoid complexity here. A big part of this handler is creating a new setting which is a different thing compared to search insight creation page logic.

valerybugakov reviewed (approved):
LGTM! 

Consider resolving Joel's comments. It's good to understand what comments were addressed when reviewing the PR, even if it's merged already.

vovakulikov commented:
@Joelkw @valerybugakov I'm about to merge this PR. All things about unifying components will be resolved when we finish the creation UI  completely, thanks for the review.
